### ISSUE TITLE
FINAL CODE

### ISSUE THREAD

**Issue by Reshma2765** [Reshma2765/DMDW#1]
import json
import numpy as np
import pandas as pd
moviesData = []

with open('./datasets/movies.json') as moviesFile:
    for line in moviesFile:
        moviesData.append(json.loads(line))

# Now take the moviesData and create a DataFrame

moviesDF = pd.DataFrame.from_records(moviesData)

# View the head of the moviesDF DataFrame

moviesDF.head()
# Let's create a dataframe with name and description
plotDF = moviesDF.loc[:, ['name', 'description']]

# Convert both columns to lowercase
plotDF.loc[:, 'name'] = plotDF.loc[:, 'name'].apply(lambda x: str(x).lower() if isinstance(x, str) else '')
plotDF.loc[:, 'description'] = plotDF.loc[:, 'description'].apply(lambda x: str(x).lower() if isinstance(x, str) else '')

# Drop all the rows which are empty
plotDF = plotDF[plotDF['description'] != '']

# Now drop duplicates in the plotDF
plotDF = plotDF.drop_duplicates()

plotDF.head()
plotDF.info()
from sklearn.feature_extraction.text import TfidfVectorizer

# Define a TF IDF Vectorizer Object
# with the removal of english stopwords turned on
tfidf = TfidfVectorizer(stop_words = 'english')

# Now costruct the TF-IDF Matrix by applying the fit_transform method on the description feature
tfidf_matrix = tfidf.fit_transform(plotDF['description'])

# View the shape of the TfIdf_matrix
tfidf_matrix.shape
tfidf_matrix
from sklearn.metrics.pairwise import linear_kernel

# Compute the cosine similarity matrix by using linear_kernel
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

# Let's view the shape of the cosine similarity matrix
print("Shape of cosine similarity matrix: ", cosine_sim.shape)

# Let's quickly view how the matrix looks like in first few column and rows
cosine_sim[0:5, 0:5]
indices = pd.Series(plotDF.index, index = plotDF['name'])
indices.head()
def plot_based_recommender(title, df = plotDF, cosine_sim = cosine_sim, indices = indices):
    # Convert title to lower-case
    title = title.lower()
    print(title)
    # Obtain the index of the movie that matched the title 
    try:
        idx = indices[title]
    except KeyError:
        print('Movie does not exist :(')
        return False

    sim_scores = list(enumerate(cosine_sim[idx]))

    # Sort the movies based on the cosine similarity scores
    sim_scores = sorted(sim_scores, key = lambda x: x[1], reverse = True)

    # Get the scores of the top 10 most similar movies. Ignore the first movie.
    sim_scores = sim_scores[1:11]

    movie_indices = [sim_score[0] for sim_score in sim_scores]

    # Return the top 10 similar movies
    return df['name'].iloc[movie_indices]   
plot_based_recommender(title='Pyaar ka Punchnama')
metaDF = moviesDF[['name', 'genre', 'language', 'director', 'cast', 'description']]
metaDF.head()
metaDF.loc[:, 'genre'] = metaDF.loc[:, 'genre'].apply(lambda x: x[:3] if len(x) > 3 else x)
metaDF.loc[:, 'cast'] = metaDF.loc[:, 'cast'].apply(lambda x: x[:3] if len(x) > 3 else x)
metaDF.loc[:, 'director'] = metaDF.loc[:, 'director'].apply(lambda x: x[:1])
metaDF.head()
def sanitize(x):
    if isinstance(x, list):
        # Strip spaces
        return [i.replace(" ", "") for i in x]
    else:
        # if it is empty, return an empty string
        if isinstance(x, str):
            return x.replace(" ", "")
        else: 
            return ''
for feature in ['director', 'cast']:
    metaDF[feature] = metaDF[feature].apply(sanitize)
metaDF.head()
def create_soup(x):
    return ' '.join(x['genre']) + ' ' + ' '.join(x['director']) + ' ' + ' '.join(x['cast'] + [x['name']] + [x['description']])

# Create the new soup feature 
metaDF['soup'] = metaDF.apply(create_soup, axis=1)

#Display the soup of the first movie 
metaDF.iloc[0]['soup']
soupDF = metaDF.loc[:, ['name', 'soup']]

# Convert both columns to lowercase
soupDF.loc[:, 'name'] = soupDF.loc[:, 'name'].apply(lambda x: x.lower())
soupDF.loc[:, 'soup'] = soupDF.loc[:, 'soup'].apply(lambda x: x.lower())

# Drop all the rows which are empty
soupDF = soupDF[soupDF['soup'] != '']

# Now drop duplicates in the soupDF
soupDF = soupDF.drop_duplicates()

soupDF.head()
print(soupDF.shape)

from sklearn.feature_extraction.text import TfidfVectorizer

# Define a TF IDF Vectorizer Object
# with the removal of english stopwords turned on
tfidf = TfidfVectorizer(stop_words = 'english')

# Now costruct the TF-IDF Matrix by applying the fit_transform method on the description feature
tfidf_matrix = tfidf.fit_transform(soupDF['soup'])

# View the shape of the TfIdf_matrix
tfidf_matrix.shape

from sklearn.metrics.pairwise import linear_kernel

# Compute the cosine similarity matrix by using linear_kernel
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

# Let's view the shape of the cosine similarity matrix
print("Shape of cosine similarity matrix: ", cosine_sim.shape)

# Let's quickly view how the matrix looks like in first few column and rows
cosine_sim[0:5, 0:5]
indices = pd.Series(soupDF.index, index = soupDF['name'])
indices.head()
def metadata_based_recommender(title, df = soupDF, cosine_sim = cosine_sim, indices = indices):
  # Convert title to lower-case
    title = title.lower()

  # Obtain the index of the movie that matched the title 
    try:
        idx = indices[title]
    except KeyError:
        print('Movie does not exist :(')
        return False
    
    sim_scores = list(enumerate(cosine_sim[idx]))

    sim_scores = sorted(sim_scores, key = lambda x: x[1], reverse = True)

    sim_scores = sim_scores[1:11]
    movie_indices = [sim_score[0] for sim_score in sim_scores]

    return df['name'].iloc[movie_indices]
metadata_based_recommender('Fan')

plot_based_recommender('Kal ho naa ho')
